upgrade_test.py

```diff
@@ -657,6 +659,83 @@ def test_rolling_upgrade(self):  # pylint: disable=too-many-locals,too-many-stat
+        stress_during_entire_upgrade = self.params.get('stress_during_entire_upgrade')
+        stress_thread_pool = self.run_stress_thread(stress_cmd=stress_during_entire_upgrade)
+        self.log.info('Sleeping for 60s to let cassandra-stress start before the rollback...')
+        time.sleep(600)
+
+        num_nodes_to_rollback = self.params.get('num_nodes_to_rollback')
+        upgraded_nodes = []
+
+        # generate random order to upgrade
+        nodes_num = len(self.db_cluster.nodes)
+        # prepare an array containing the indexes
+        indexes = list(range(nodes_num))
+        # shuffle it so we will upgrade the nodes in a random order
+        random.shuffle(indexes)
+
+        upgrade_sstables = self.params.get('upgrade_sstables')
+
+        if num_nodes_to_rollback > 0:
+            # Upgrade all nodes that should be rollback later
+            for i in range(num_nodes_to_rollback):
+                InfoEvent(f'Step{i + 1} - Upgrade node{i + 1}')
+                self.db_cluster.node_to_upgrade = self.db_cluster.nodes[indexes[i]]
+                self.log.info(f'Upgrade Node {self.db_cluster.node_to_upgrade.name} begins')
+                self.upgrade_node(self.db_cluster.node_to_upgrade, upgrade_sstables=upgrade_sstables)
+                self.log.info(f'Upgrade Node {self.db_cluster.node_to_upgrade.name} ended')
+                self.db_cluster.node_to_upgrade.check_node_health()
+                upgraded_nodes.append(self.db_cluster.node_to_upgrade)
+
+            # Rollback all nodes that where upgraded (not necessarily in the same order)
+            random.shuffle(upgraded_nodes)
+            self.log.info(f'Upgraded Nodes to be rollback are: {upgraded_nodes}')
+            for node in upgraded_nodes:
+                InfoEvent(f'Step{num_nodes_to_rollback + upgraded_nodes.index(node) + 1} - '
+                          f'Rollback node{upgraded_nodes.index(node) + 1}')
+                self.log.info(f'Rollback Node {node} begin')
+                self.rollback_node(node, upgrade_sstables=upgrade_sstables)
+                self.log.info(f'Rollback Node {node} ended')
+                self.db_cluster.node.check_node_health()
+
+        # Upgrade all nodes
+        for i in range(nodes_num):
+            InfoEvent(f'Step{num_nodes_to_rollback * 2 + i + 1} - Upgrade node{i + 1}')
+            self.db_cluster.node_to_upgrade = self.db_cluster.nodes[indexes[i]]
+            self.log.info(f'Upgrade Node {self.db_cluster.node_to_upgrade.name} begins')
+            self.upgrade_node(self.db_cluster.node_to_upgrade, upgrade_sstables=upgrade_sstables)
+            self.log.info(f'Upgrade Node {self.db_cluster.node_to_upgrade.name} ended')
+            self.db_cluster.node_to_upgrade.check_node_health()
+            upgraded_nodes.append(self.db_cluster.node_to_upgrade)
+
+        InfoEvent(f'All nodes were upgraded successfully')
+
+        InfoEvent('Waiting for stress_during_entire_upgrade to finish')
```

Comment:
Why not combine the two events into one event?

Reply:
Two separate logics.